design/034-new_lockfiles.md

```diff
@@ -0,0 +1,237 @@
+
+All of that resulted in lockfiles that have been able to partially address some of the related problems, but that in general are complex to use, understand, and implement CI with them.
+
+So, for Conan 2.0, we are proposing a way more “traditional” approach to lockfiles, with the simplifications enumerated above, making Conan lockfiles something closer to what other package managers like NPM can do. There were a few unknowns if this was doable, so for this proposal we wanted to implement it first, check below for details.
+
+## Detailed Design
+
+In the surface, the basic usage of lockfiles hasn’t change much, creating and using a lockfile would be:
+
+Lets say we have conanfile.txt:
+
+```
+[requires]
+pkg/[>=0.0 <1.0]
+```
+
+And existing ``pkg/0.1`` then:
+
+```bash
+conan lock create .  --lockfile-out=conan.lock
+# this will create a lockfile with a “requires” list containing “pkg/0.1”
+# no matter if other new version pkg/0.2 in the range is created now
+# this will keep resolving to “pkg/0.1”
+conan install . --lockfile=conan.lock
+```
+
+But many things have changed, lets see them.
+
+### Non-strict vs strict
+
+If we now do a modification to our conanfile.txt to:
+
+```
+[requires]
+pkg/[>=1.0 <2.0]
+```
+
+And assuming a ``pkg/1.1`` version exist, then:
+
+```bash
+conan install . --lockfile=conan.lock  # will work
+```
+
+Will not fail by default, and will be able to resolve to ``pkg1.1``. This is because the default mode is non-strict. One of the most frustrating behaviors of Conan 1.X lockfiles was when lockfiles didn’t allow you to do modifications to your conanfiles and keep working.
+
+If we do:
+
+```bash
+conan install . --lockfile=conan.lock --lockfile-strict # will not work
+```
+
+It will throw an error saying that the requirement ``pkg/[>=1.0 <2.0]`` cannot be found in the lockfile
+
+### Multi-configuration lockfiles
+
+If we have one conanfile like this:
+
+```python
+class Pkg(ConanFile):
+    settings = "os"
+    def requirements(self):
+        if self.settings.os == "Windows":
+            self.requires("win/[>0.0]")
+        else:
+            self.requires("nix/[>0.0]")
+```
+Then it will be possible to capture the lockfile for both Windows and Linux configuration in a single lockfile file, in the following way, assuming there exist packages ``win/0.1`` and ``nix/0.1``:
+
+```bash
+conan lock create .  --lockfile-out=conan.lock -s os=Windows -s:b os=Windows
+# That captures a lockfile with ``win/0.1``
+conan lock create .  --lockfile=conan.lock --lockfile-out=conan.lock -s os=Linux -s:b os=Linux
+# That will augment the conan lock and it will end with both ``win/0.1`` and ``nix/0.1``
```

Comment:
It would be great if it were possible to do both that in a single conan invocation, at least with profiles.

Something like:

```
conan lock create . --lockfile-out=conan.lock --profile windows-profile --profile linux-profile --profile ios-profile --profile android-arm64-profile --profile emscripten --profile macos-profile ...
```

Reply:
This will be easily implementable with the custom commands and the public API at your will.

Note that your example, for a start is already having a different behavior, as multiple ``--profile`` arguments in CLI actually compound and collapse into a single one. So a new syntax would be necessary to define such different profiles. Multiply that by 2 (host and build profiles), and by 3 for (settings, options, conf, that are also part of the profiles), and we have made the ``conan lock create`` CLI arguments a nightmare, that would probably not satisfy many users.

But with the custom commands you can perfectly define what you want, what arguments you will be providing, the behavior that you want, and using the Conan Public API, you will be able to code it in relatively few lines (and distribute your custom commands with ``conan config install`` too)